src/infer.c

```diff
@@ -0,0 +1,333 @@
+/*
+* Copyright 2021 The SELint Contributors
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "infer.h"
+#include "color.h"
+#include "maps.h"
+#include "util.h"
+
+/* Uses directly in the testsuite */
+enum selint_error infer_interfaces_shallow(const struct policy_node *node);
+enum selint_error infer_interfaces_deep(const struct policy_node *node);
+
+static enum param_flavor name_to_param_flavor(enum name_flavor flavor)
+{
+	switch (flavor) {
+	case NAME_TYPE:
+		return PARAM_TYPE;
+	case NAME_TYPEATTRIBUTE:
+		return PARAM_TYPEATTRIBUTE;
+	case NAME_TYPE_OR_ATTRIBUTE:
+		return PARAM_TYPE_OR_ATTRIBUTE;
+	case NAME_ROLE:
+		return PARAM_ROLE;
+	case NAME_ROLEATTRIBUTE:
+		return PARAM_ROLEATTRIBUTE;
+	case NAME_ROLE_OR_ATTRIBUTE:
+		return PARAM_ROLE_OR_ATTRIBUTE;
+	case NAME_CLASS:
+		return PARAM_CLASS;
+	case NAME_OBJECT_NAME:
+		return PARAM_OBJECT_NAME;
+	default:
+		// should never happen
+		return PARAM_UNKNOWN;
+	}
+}
+
+enum infer_type { IN_SHALLOW, IN_DEEP };
+
+struct infer_data {
+	struct interface_trait *if_data;
+	enum infer_type mode;
+	const struct policy_node *node;
+};
+
+static const char *trait_type_to_str(enum trait_type t)
+{
+	switch (t) {
+	case INTERFACE_TRAIT:
+		return "interface";
+	case TEMPLATE_TRAIT:
+		return "template";
+	case MACRO_TRAIT:
+		return "macro";
+	default:
+		// should never happen
+		return "unknown-trait-type";
+	}
+}
+
+static void infer_func(const char *name, enum name_flavor flavor, unsigned short id, void *visitor_data)
```

Comment:
Can this (and infer_interface) get a comment explaining their usage?  It's not clear to me from the names, and it's slow progress trying to infer from the implementation.

Reply:
Okay, reading more, I *think* I have a rough idea of what these are both for based on how they're used.  I still would like to have something with the functions to help orient anyone reading the code.